src/hashfunctions.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ *
+ * hashfun: various hash functions
+ */
+
+#ifndef HASHFUNCTIONS_H
+#define HASHFUNCTIONS_H
+
+#include "datastructures.h"
+#include "src/intfuncs.h"
+
+extern struct DatastructuresModule HashFunctionsModule;
+
+
+// Hash two integers together
+static inline UInt HashCombine2(UInt hash1, UInt hash2);
+
+// Transform a UInt into a GAP intermediate integer, shrinking
+// the size of the number as required
+static inline Obj HashValueToObjInt(UInt hash);
+
+// We do not use INTOBJ_INT here, because our UInt might not
+// fit inside an INTOBJ, but we do not care about throwing the
+// top bits away
+static inline Obj HashValueToObjInt(UInt hash)
+{
+    return (Obj)((hash << 2) + 0x01);
+}
+
+// Perform a shuffle of a UInt. Ideally, changing any
+// bit would have a 50/50 chance of changing every other bit.
+// The main purpose of this is to allow adding values when
+// we want to hash an unordered list -- while 1+3 = 2+2,
+// HashUInt(1) + HashUInt(3) should not be equal to
+// HashUInt(2) + HashUInt(2)
+// From: http://www.concentric.net/~Ttwang/tech/inthash.htm
+static UInt ShuffleUInt(UInt key);
+
+#ifdef SYS_IS_64_BIT
+
+// TODO: Make better!
```

Comment:
"Make better" in which way? Either remove this comment, or else add a (brief) explanation as to what needs to be better.